compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CallableReferenceLowering.kt

```diff
@@ -32,140 +20,151 @@ import org.jetbrains.kotlin.backend.jvm.codegen.isInlineIrExpression
+            !expression.symbol.owner.parentAsClass.defaultType.isFunctionOrKFunction())
+            return expression
+
+        return IrCallImpl(
+            expression.startOffset, expression.endOffset,
+            expression.type, functionNInvokeFun.symbol, functionNInvokeFun.descriptor,
+            1, expression.origin
+        ).apply {
+            putTypeArgument(0, expression.type)
+            dispatchReceiver = expression.dispatchReceiver
+            extensionReceiver = expression.extensionReceiver
+            val vararg = IrVarargImpl(
+                UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                context.ir.symbols.array.typeWith(context.irBuiltIns.anyNType),
+                context.irBuiltIns.anyNType,
+                (0 until expression.valueArgumentsCount).map { expression.getValueArgument(it)!! }
+            )
+            putValueArgument(0, vararg)
+        }
+    }
 
-                //TODO: clean
-                return super.visitFunctionAccess(newCall)
-            }
+    private fun markInlineFunctionReferences(expression: IrFunctionAccessExpression) {
+        val function = expression.symbol.owner
+        if (!function.isInlineFunctionCall(context))
+            return
 
-            override fun visitFunctionReference(expression: IrFunctionReference): IrExpression {
-                expression.transformChildrenVoid(this)
+        for (parameter in function.valueParameters) {
+            if (!parameter.isInlineParameter())
+                continue
 
-                if (!expression.type.toKotlinType().isFunctionOrKFunctionType || inlineLambdaReferences.contains(expression)) {
-                    // Not a subject of this lowering.
-                    return expression
-                }
+            val valueArgument = expression.getValueArgument(parameter.index) ?: continue
+            if (!isInlineIrExpression(valueArgument))
+                continue
 
-                val currentDeclarationParent = allScopes.map { it.irElement }.last { it is IrDeclarationParent } as IrDeclarationParent
-                val loweredFunctionReference = FunctionReferenceBuilder(currentDeclarationParent, expression).build()
-                return context.createIrBuilder(currentScope!!.scope.scopeOwnerSymbol).irBlock(expression) {
-                    +loweredFunctionReference.functionReferenceClass
-                    +irCall(loweredFunctionReference.functionReferenceConstructor.symbol).apply {
-                        expression.getArguments().forEachIndexed { index, argument ->
-                            putValueArgument(index, argument.second)
-                        }
-                    }
-                }
+            if (valueArgument is IrFunctionReference) {
+                ignoredFunctionReferences.add(valueArgument)
+            } else if (valueArgument is IrBlock) {
+                ignoredFunctionReferences.addIfNotNull(valueArgument.statements.firstIsInstanceOrNull())
```

Comment:
Actually it should be `singletIsInstanceOrNull`

Reply:
And wouldn't that be a nice function to have in the standard library? :-)

Fixed!